### PR TITLE
[Driver][SYCL][NewOffload] Fix arch settings for nvptx and amd

### PR DESCRIPTION
**Pull request by mdtoguchi**
When compiling for -fsycl-targets values of nvptx64-nvidia-cuda and amdgcn-amd-gpu, the default arch behaviors were not applied to the compilation.  Updates to do the following:
 - Add default of sm_50 for nvptx64 if not provided
 - Emit diagnostic if no arch provided for amd
 - Parse -Xsycl-backend-target for offload-arch values

### CODE REVIEW

**Review by asudarsa — APPROVED** [review]
I tested the changes for the new offloading model on several failing SYCL E2E tests on an AMD machine and the fix resolves all of those failures. Test added looks fine.

Thanks much for the fix.

File: clang/lib/Driver/Driver.cpp [hunk 1]
```diff
@@ -7851,8 +7867,18 @@ Driver::getOffloadArchs(Compilation &C, const llvm::opt::DerivedArgList &Args,
       Archs.insert(CudaArchToString(CudaArch::HIPDefault));
     else if (Kind == Action::OFK_OpenMP)
       Archs.insert(StringRef());
-    else if (Kind == Action::OFK_SYCL)
-      Archs.insert(StringRef());
+    else if (Kind == Action::OFK_SYCL) {
+      // For SYCL offloading, we need to check the triple for NVPTX or AMDGPU
+      // and set the default arch accordingly
+      // mtoguchi
```
Comment (hchilama) [inline clang/lib/Driver/Driver.cpp:7873]: Can you remove this?

Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp]: Done

**Review by hchilama — APPROVED** [review]

**Comment by mdtoguchi** [comment]
@intel/llvm-gatekeepers, This should be ready for merge - please take a look.

File: clang/lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -7851,8 +7867,17 @@ Driver::getOffloadArchs(Compilation &C, const llvm::opt::DerivedArgList &Args,
...
       Archs.insert(StringRef());
-    else if (Kind == Action::OFK_SYCL)
-      Archs.insert(StringRef());
+    else if (Kind == Action::OFK_SYCL) {
+      // For SYCL offloading, we need to check the triple for NVPTX or AMDGPU
+      // and set the default arch accordingly
+      if (TC->getTriple().isNVPTX())
+        Archs.insert(CudaArchToString(CudaArch::SM_50));
+      else if (TC->getTriple().isAMDGPU())
+        C.getDriver().Diag(clang::diag::err_drv_sycl_missing_amdgpu_arch)
```
Comment (srividya-sundaram) [inline clang/lib/Driver/Driver.cpp:7876]: >> we need to check the triple for NVPTX or AMDGPU and set the default arch accordingly
Why are we just checking if the triple is AMD and just emitting an error? Shouldn't we set the arch first?

Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp:7876]: The error is that there is a missing arch setting.  The AMD path requires an arch to be supplied (unlike nvptx where we set a default arch).

Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp:7876]: I will update the description to be more clear of our intentions in regards to AMD

Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp:7877]: I have updated the comment.

**Comment by asudarsa** [comment]
We should really be not running CI testing for comment updates...:-(

**Comment by martygrant** [comment]
@mdtoguchi this is awaiting approval from @srividya-sundaram

**Comment by asudarsa** [comment]
We have approval from a driver team reviewer, This change unblocks development by fixing a large number of test failures. So, I ill merge it now. @srividya-sundaram, please feel free to reach out to me or @mdtoguchi for any post-commit changes if required.

Thanks
Sincerely

**Comment by srividya-sundaram** [comment]
Sorry, I missed the notification for approval request.

**Pull request merged**